src/ott/neural/solvers/expectile_neural_dual.py

```diff
@@ -0,0 +1,655 @@
+    """Transport the points from source to the target distribution."""
+    source = jnp.atleast_2d(source)
+    assert (forward
+            is True), ("Only forward mapping (source -> target) is supported.")
+    return self._grad_f(source)
+
+
+class EuclideanPotentials(DualPotentials):
+  r"""The dual potentials, trained with scalar product operation.
+
+  Args:
+  g: The second dual potential function.
+  grad_f: Gradient of the first dual potential function.
+  """
+
+  def __init__(self, g: PotentialValueFn_t, grad_f: PotentialGradientFn_t):
+
+    def g_conjugate(x: jnp.ndarray) -> jnp.ndarray:
+      y_hat = jax.lax.stop_gradient(grad_f(x))
+      return -g(y_hat) + jnp.dot(x, y_hat)
+
+    super().__init__(g_conjugate, g, cost_fn=costs.SqEuclidean(), corr=True)
+    self.__grad_f = grad_f
+
+  @property
+  def _grad_f(self) -> Callable[[jnp.ndarray], jnp.ndarray]:
+    return jax.vmap(self.__grad_f)
+
+  def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport from source to the target distribution or vice-versa."""
+    vec = jnp.atleast_2d(vec)
+    return self._grad_f(vec) if forward else self._grad_g(vec)
+
+  def distance(self, src: jnp.ndarray, tgt: jnp.ndarray) -> float:
+    """W2 distance."""
+    src, tgt = jnp.atleast_2d(src), jnp.atleast_2d(tgt)
+    f = jax.vmap(self.f)
+    g = jax.vmap(self.g)
+    corr = jnp.mean(f(src)) + jnp.mean(g(tgt))
+    return -2.0 * corr + jnp.mean(jnp.sum(src ** 2, axis=-1)) + \
+      jnp.mean(jnp.sum(tgt ** 2, axis=-1))
+
+
+def expectile_loss(
+    adv: jnp.ndarray, diff: jnp.ndarray, expectile: float = 0.9
+) -> jnp.ndarray:
+  weight = jnp.where(adv >= 0, expectile, (1 - expectile))
+  return weight * diff ** 2
+
+
+@jax.tree_util.register_pytree_node_class
+class DotCost(costs.CostFn):
+  """The cost function that is used in OT training with scalar product."""
+
+  def pairwise(self, x: jnp.ndarray, y: jnp.ndarray) -> float:
+    """Dot product of vectors."""
+    return jnp.dot(x, y)
+
+  @classmethod
+  def _padder(cls, dim: int) -> jnp.ndarray:
+    return jnp.ones((1, dim))
+
+
+class PotentialTrainState(train_state.TrainState):
+  r"""Adds information about the model's value and gradient to the state.
+
+  The gradient may differ from composition of jax.grad and value functions.
+
+  Args:
+  potential_value_fn: the potential's value function.
+  potential_gradient_fn: the potential's gradient function.
+  """
+
+  potential_value_fn: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                               PotentialValueFn_t] = struct.field(
+                                   pytree_node=False
+                               )
+
+  potential_gradient_fn: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                                  PotentialGradientFn_t] = struct.field(
+                                      pytree_node=False
+                                  )
+
+
+class PotentialModelWrapper(nn.Module):
```

Comment:
Might be worthwhile subclassing `ott.neural.networks.potentials.BasePotential` and overriding `potential_value_fn` instead.

Reply:
In doing so, we get `potential_gradient_fn` for free, as well as `create_train_state`.